feature: new github action which limit the file changes in single PR to 20

Tarunmeena0901 opened the pull request:
<!--
This section can be deleted after reading.

We employ the following branching strategy to simplify the development process and to ensure that only stable code is pushed to the `master` branch:

- `develop`: For unstable code: New features and bug fixes.
- `master`: Where the stable production ready code lies. Only security related bugs.

NOTE!!!

ONLY SUBMIT PRS AGAINST OUR `DEVELOP` BRANCH. THE DEFAULT IS `MAIN`, SO YOU WILL HAVE TO MODIFY THIS BEFORE SUBMITTING YOUR PR FOR REVIEW. PRS MADE AGAINST `MAIN` WILL BE CLOSED.
-->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**
Feature
<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Issue Number:**

Fixes #1276 <!--Add related issue number here.-->

**Did you add tests for your changes?**

<!--Yes or No. Note: Add unit tests or automation tests for your code.-->

**Snapshots/Videos:**
![Screenshot 2023-12-29 014258](https://github.com/PalisadoesFoundation/talawa-admin/assets/97682967/40be5ebd-ffbc-4d19-a90e-a6ec5ee5587a)

![Screenshot 2023-12-29 014152](https://github.com/PalisadoesFoundation/talawa-admin/assets/97682967/da2d64d5-c3f8-4dba-9b6c-a95db98c2f71)

<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**

<!--Add link to Talawa-Docs.-->

**Summary**

- new contributors were creating PRs from branch whose source was not our `develop` branch which was causing PRs with hundreds of submitted files.

- To solve this issue I added a new github action which count the number of file-changes made in a single PR and if the number of changed file exceeds 20 it give user error message "Contributor may be merging into an incorrect branch"

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**

- new contributors who are not very much aware of working of github will have a hint to fix this problem everytime they make a wrong PR 
- This change encourages contributors to align with project practices, reducing the likelihood of unintentional merges into incorrect branches. 

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Changes Made:**
- I added a python script `check-changed-files.py` and modified the `pull-request.yml` file to add the required action
- The `Check-Changed-File` job is dependent on `Code-Quality-Check` job so it will only run if `Code-Quality-Check` is successfull.
- python `black` and `pydocstyle` linting and formatting rules are applied on every PR

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**
YESSSSS
<!--Yes or No-->

palisadoes reviewed (changes requested):
For the python script:

1. Follow the Google Python style guide especially in the DocString area. https://google.github.io/styleguide/pyguide.html
2. Make sure that the script is python `black`, `pydocstring`, `pylint`, `flake8` and `pycodestyle` compliant
4. Use the `countline.py` script in the `workflow` directory as a guide

![image](https://github.com/PalisadoesFoundation/talawa-admin/assets/16875803/eee20105-1c68-467d-b1cb-9371240a0678)

Tarunmeena0901 commented:
Thanks for noticing , I made the python script  `black`, `pydocstring`, `pylint`, `flake8` and `pycodestyle` compliant 
also changed the name of script to a more descriptive name

palisadoes commented:
Please work with @noman2002 on this. He'll be taking over the review of this PR

Tarunmeena0901 commented:
Oh ok thanks for your time

noman2002 reviewed (dismissed):
Looks good to me. But let someone else also review it.

Tarunmeena0901 commented:
Yeah sure , no hurry 🤞🏻 I'll be glad to hear some suggestions

palisadoes commented:
1. Please fix the failing test.
2. It's failing when only two files were updated

Tarunmeena0901 commented:
my bad there was a small typo in `.yml file` i fixed it

palisadoes commented on .github/workflows/pull-requests.yml:
```diff
@@ -52,6 +52,31 @@ jobs:
+        with:
+          python-version: 3.9
+
+      - name: Install dependencies
```
Remove references to python black

palisadoes commented on .github/workflows/pull-requests.yml:
```diff
@@ -52,6 +52,31 @@ jobs:
+        run: |
+          python .github/workflows/count_changed_files.py "${{ github.base_ref }}" "${{ github.head_ref }}"
+
+      - name: Run Black
```
Remove references to python black

palisadoes reviewed (changes requested):
1. See comments
2. Remove references to python black. It is causing the workflow to fail

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,103 @@
+        base_branch = f"origin/{base_branch}"
+        current_branch = f"origin/{current_branch}"
+        # Run git command to get the list of changed files
+        command = f"git diff --name-only {base_branch}...{current_branch} | wc -l"
```
1. Use `subprocess.Popen.communicate`. 
1. It will give you better error messaging if the command fails. 
    - https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,103 @@
+    try:
+        # Get base branch and current commit
+        base_branch = sys.argv[1]
+        print(base_branch)
```
If this is informational, then there must be more explicatory status text, otherwise remove it.

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,103 @@
+        base_branch = sys.argv[1]
+        print(base_branch)
+        current_branch = sys.argv[2]
+        print(current_branch)
```
If this is informational, then there must be more explicatory status text, otherwise remove it.

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,103 @@
+    """
+    try:
+        base_branch = f"origin/{base_branch}"
+        current_branch = f"origin/{current_branch}"
```
Only one statement in the `try:` block. It makes troubleshooting easier

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,103 @@
+    """
+    try:
+        # Get base branch and current commit
+        base_branch = sys.argv[1]
```
Only one statement per `try:` block. It makes troubleshooting easier

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,103 @@
+        # Get base branch and current commit
+        base_branch = sys.argv[1]
+        print(base_branch)
+        current_branch = sys.argv[2]
```
Use argparse for passing arguments. It makes the code more readable and it gives you a lot of argument parsing flexibility such as specifying the type of input expected and whether the argument is mandatory or not.

palisadoes reviewed (changes requested):
See comments

Tarunmeena0901 commented:
thanks for giving me info about `argparser` and `subprocess.communicate` I did not knew about these . i made the suggested changes and tested them as well.
1. only one statement per try block for better troubleshooting
2. used argparse for passing arguments
3. added some print statement regarding on and from which branch one is creating a pull request
4. removed the `black` reference from .yml file

![Screenshot 2024-01-02 161301](https://github.com/PalisadoesFoundation/talawa-admin/assets/97682967/46dbd675-0cc4-470c-a676-e5ede794f981)

palisadoes reviewed (changes requested):
See comments

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,135 @@
+    file_count = _count_changed_files(base_branch, current_branch)
+    print(f"Number of changed files: {file_count}")
+    # Check if the count exceeds 20
+    if file_count > 20:
```
Get this value from argparse with a default of `20`

Tarunmeena0901 replied:
done

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,135 @@
+        "base_branch", type=str, help="Base branch where pull request should be made."
+    ),
+    parser.add_argument(
+        "current_commit",
```
Shouldn't this string refer to the branch of the PR? This implies a commit ID. Please change.

Tarunmeena0901 replied:
this string refers to branch from which PR is being made on the contributor side

palisadoes reviewed (changes requested):
See comments

Tarunmeena0901 commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,135 @@
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "base_branch", type=str, help="Base branch where pull request should be made."
```
well its not mandatory may be we can remove it cause anyone working with this file would already know that base branch can be a name or SHA string

Tarunmeena0901 replied:
so implies in current_commit argument

Tarunmeena0901 commented on .github/workflows/pull-requests.yml:
```diff
@@ -52,6 +52,23 @@ jobs:
+
+      - name: Run Python script
+        run: |
+          python .github/workflows/count_changed_files.py "${{ github.base_ref }}" "${{ github.head_ref }}"
```
ok sure

Tarunmeena0901 commented:
thanks for pointing out my mistakes and suggesting changes I made following changes in my most recent commit :

- removed unnecessary try blocks

- in `parser.arguments` remove not mandatory arguments like`type = str`

- changed the name of variable which represent contributors PR branch to more self describable
   `current_commit ==> PR_branch`

- add the parameters to make it obvious what's happening 
  `python .github/workflows/count_changed_files.py --base_branch "${{ github.base_ref }}" --PR_branch "${{ github.head_ref }}"`

- removed unnecessary library imports like os

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,129 @@
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--base_branch",
+        help="Base branch where pull request should be made."
```
Make this required, and type string

Tarunmeena0901 replied:
done

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,129 @@
+        help="Base branch where pull request should be made."
+    ),
+    parser.add_argument(
+        "--PR_branch",
```
Make this required, and type string

Tarunmeena0901 replied:
done

palisadoes reviewed (changes requested):
See comments

Tarunmeena0901 commented:
I made the changes sir , thanks for your constant feedback

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,133 @@
+        help="Base branch where pull request should be made."
+    ),
+    parser.add_argument(
+        "--PR_branch",
```
Make this lowercase throughout for consistency. Python prefers lowercase variable names.

Tarunmeena0901 replied:
done sir

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,133 @@
+        )
+        output, error = process.communicate()
+
+        if process.returncode != 0:
```
1. Move this and the following statements out of the `try:` block.  
2. Only use one statement per `try:` block, and only for statements where success is not guaranteed.

Tarunmeena0901 replied:
oh ok i will keep that in mind

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,133 @@
+        process = subprocess.Popen(
+            command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        output, error = process.communicate()
```
Only have this statement in the `try` block. It is the only one where success is not guaranteed.

palisadoes replied:
When you have more than one statement in the block, troubleshooting becomes harder

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,133 @@
+    args = _arg_parser_resolver()
+    base_branch = args.base_branch
+    print(f"You are trying to merge on branch: {base_branch}")  # Print for verification
+    PR_branch = args.PR_branch
```
Define this variable near the `args =` definition. It makes the code more readable

Tarunmeena0901 replied:
yeah I think that would be nice

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,133 @@
+    # Count changed files
+    file_count = _count_changed_files(base_branch, PR_branch)
+    print(f"Number of changed files: {file_count}")
+    # Check if the count exceeds 20
```
Add space above. Readability

Tarunmeena0901 replied:
done

palisadoes commented on .github/workflows/count_changed_files.py:
```diff
@@ -0,0 +1,133 @@
+    print(
+        f"You are making commit from your branch: {PR_branch}"
+    )  # Print for verification
+    # Count changed files
```
Add space above. Readability

Tarunmeena0901 replied:
done

palisadoes reviewed (changes requested):
See comments. This should will be ready to merge very soon.

Tarunmeena0901 commented:
> See comments. This should will be ready to merge very soon.

that sounds so good 🤩 , thank you for your constant guidance through out the issue

- I made the changes you suggested and tested it locally on my own repo and its working perfectly 
- please suggest if you notice any more valid changes need to be done

Tarunmeena0901 commented:
I just saw a PR #1317 where number of file changes my 23 these type of PR wont pass from git checks if merge the git hub action from this PR . should we increase the default value of number of file changes allowed to 30 currently its 20 ?

palisadoes commented:
> I just saw a PR #1317 where number of file changes my 23 these type of PR wont pass from git checks if merge the git hub action from this PR . should we increase the default value of number of file changes allowed to 30 currently its 20 ?

This is meant to be a guide for reviewers

palisadoes commented:
We can adjust the number afterwards

Tarunmeena0901 commented:
#1276
